guides/0010-defining-a-pipeline.md

```diff
@@ -0,0 +1,111 @@
+# Defining a Pipeline
+
+Astarte Flow has its own DSL for pipeline definition.
+
+# Pipe Operator
+
+The core construct behind Astarte Flow pipelines is the pipe operator `|` which connects two
+blocks.
+
+In the following example messages flowing out from block1 (the source) will be processed from
+block2 and they will be processes from block3 (the sink):
+```
+    block1 | block2 | block3
+```
+
+New lines are allowed (and suggested) for readabilty reasons.
+
+# With Operator
+
+Most of the times blocks are not usable alone without any kind of configuration.
+
+Astarte Flow DSL defines a `.` operator that can be read as "with". Several `.` can be chained
+together.
+
+Syntax for `.` operator is `block.property1(value1).property2(value2)`.
+
+The following example can be read as: "Pipe messages from `astarte_devices_source` block with
+`realm` `"test"` to `filter` block with script `"return message.value > 0"`.
+
+```
+astarte_devices_source
+  .realm("test")
+| filter
+  .script("return message.value > 0")
+```
+
+# Literals
+
+In Astarte Flow DSL literals are meant to be used with `.` operator.
+
+## Integers
+
+Base 10 integers are allowed, such as `-1`, `0` and `89`.
+
+## Booleans
+
+`true` and `false` boolean values are allowed.
+
+## Strings
+
+Strings are marked with `""`, such as `"a string"`.
+In some situations it might be convenient using multi-line strings such as:
+```
+"""
+A
+"multi-line"
+string
+"""
+```
+
+Multi-line strings are convenient for Lua scripting.
+
+## Objects
+
+JSON like objects are allowed, however there is a major difference: `""` around keys are not
+required.
```

Comment:
Saying "`""` around keys are not _required_" is actually wrong because if you add double quotes the parses interprets the config value as a string. This should become:

JSON objects and arrays are required for the configuration of some blocks and they must be provided without wrapping them in a string

```
virtual_device_pool
    .realms([{"realm": "foo", "jwt": "bar"}]
```